File: libpretixsync/src/main/java/eu/pretix/libpretixsync/sync/AllEventsSyncAdapter.java
package eu.pretix.libpretixsync.sync;

import eu.pretix.libpretixsync.api.PretixApi;
import eu.pretix.libpretixsync.db.Event;
import io.requery.BlockingEntityStore;
import io.requery.Persistable;
import io.requery.query.Tuple;
import io.requery.util.CloseableIterator;

import org.joda.time.format.ISODateTimeFormat;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.Iterator;
import java.util.List;

public class AllEventsSyncAdapter extends BaseDownloadSyncAdapter<Event, String> {
    public AllEventsSyncAdapter(BlockingEntityStore<Persistable> store, FileStorage fileStorage, PretixApi api, String syncCycleId, SyncManager.ProgressFeedback feedback) {
        super(store, fileStorage, "__all__", api, syncCycleId, feedback);
    }

    @Override
    CloseableIterator<Tuple> getKnownIDsIterator() {
        return store.select(Event.SLUG)
                .get().iterator();
    }

    @Override
    public void updateObject(Event obj, JSONObject jsonobj) throws JSONException {
        obj.setSlug(jsonobj.getString("slug"));
        obj.setCurrency(jsonobj.getString("currency"));
        obj.setDate_from(ISODateTimeFormat.dateTimeParser().parseDateTime(jsonobj.getString("date_from")).toDate());
        if (!jsonobj.isNull("date_to")) {
            obj.setDate_to(ISODateTimeFormat.dateTimeParser().parseDateTime(jsonobj.getString("date_to")).toDate());
        }
        obj.setLive(jsonobj.getBoolean("live"));
        obj.setHas_subevents(jsonobj.getBoolean("has_subevents"));
        obj.setJson_data(jsonobj.toString());
    }

    @Override
    protected String getUrl() {
        return api.organizerResourceUrl(getResourceName());
    }

    @Override
    String getResourceName() {
        return "events";
    }

    @Override
    String getId(JSONObject obj) throws JSONException {
        return obj.getString("slug");
    }

    @Override
    String getId(Event obj) {
        return obj.getSlug();
    }

    @Override
    Event newEmptyObject() {
        return new Event();
    }

    @Override
    public CloseableIterator<Event> runBatch(List<String> parameterBatch) {
        return store.select(Event.class)
                .where(Event.SLUG.in(parameterBatch))
                .get().iterator();
    }
}

File: libpretixsync/src/main/java/eu/pretix/libpretixsync/sync/AllEventsSyncAdapter.kt
package eu.pretix.libpretixsync.sync

import app.cash.sqldelight.TransactionWithoutReturn
import app.cash.sqldelight.db.QueryResult
import eu.pretix.libpretixsync.api.PretixApi
import eu.pretix.libpretixsync.sqldelight.Event
import eu.pretix.libpretixsync.sqldelight.SyncDatabase
import eu.pretix.libpretixsync.sync.SyncManager.ProgressFeedback
import org.joda.time.format.ISODateTimeFormat
import org.json.JSONObject

class AllEventsSyncAdapter(
    db: SyncDatabase,
    fileStorage: FileStorage,
    api: PretixApi,
    syncCycleId: String,
    feedback: ProgressFeedback?,
) : SqBaseDownloadSyncAdapter<Event, String>(
    db = db,
    api = api,
    syncCycleId = syncCycleId,
    eventSlug = "__all__",
    fileStorage = fileStorage,
    feedback = feedback,
) {
    override fun getResourceName(): String = "events"

    override fun getUrl(): String {
        return api.organizerResourceUrl(getResourceName())
    }

    override fun getId(obj: Event): String = obj.slug!!

    override fun getId(obj: JSONObject): String = obj.getString("slug")

    override fun getJSON(obj: Event): JSONObject = JSONObject(obj.json_data!!)

    override fun queryKnownIDs(): MutableSet<String>? {
        val res = mutableSetOf<String>()
        db.eventQueries.selectSlugs().execute { cursor ->
            while (cursor.next().value) {
                val id = cursor.getString(0) ?: throw RuntimeException("slug column not available")
                res.add(id)
            }

            QueryResult.Unit
        }

        return res
    }

    override fun insert(jsonobj: JSONObject) {
        val dateFrom =
            ISODateTimeFormat.dateTimeParser().parseDateTime(jsonobj.getString("date_from"))
                .toDate()

        val dateTo = if (!jsonobj.isNull("date_to")) {
            ISODateTimeFormat.dateTimeParser().parseDateTime(jsonobj.getString("date_to"))
                .toDate()
        } else {
            null
        }

        db.eventQueries.insert(
            currency = jsonobj.getString("currency"),
            date_from = dateFrom,
            date_to = dateTo,
            has_subevents = jsonobj.getBoolean("has_subevents"),
            json_data = jsonobj.toString(),
            live = jsonobj.getBoolean("live"),
            slug = jsonobj.getString("slug"),
        )
    }

    override fun update(obj: Event, jsonobj: JSONObject) {
        val dateFrom =
            ISODateTimeFormat.dateTimeParser().parseDateTime(jsonobj.getString("date_from"))
                .toDate()

        val dateTo = if (!jsonobj.isNull("date_to")) {
            ISODateTimeFormat.dateTimeParser().parseDateTime(jsonobj.getString("date_to")).toDate()
        } else {
            null
        }

        db.eventQueries.updateFromJson(
            currency = jsonobj.getString("currency"),
            date_to = dateTo,
            date_from = dateFrom,
            has_subevents = jsonobj.getBoolean("has_subevents"),
            json_data = jsonobj.toString(),
            live = jsonobj.getBoolean("live"),
            slug = obj.slug,
        )
    }

    override fun delete(key: String) {
        db.eventQueries.deleteBySlug(key)
    }

    override fun runInTransaction(body: TransactionWithoutReturn.() -> Unit) {
        db.eventQueries.transaction(false, body)
    }

    override fun runBatch(parameterBatch: List<String>): List<Event> =
        db.eventQueries.selectBySlugList(parameterBatch).executeAsList()
}
